### PR TITLE
Issue 2854: (SegmentStore) Segment Container Extensions

### LINKED ISSUE #2854: Segment Container Extensions
**Problem description**
We need the ability to load up external code (extensions) into the core of the Segment Store without having to modify the code of the Segment Store Core. These extensions will have access to lower-level Segment Container APIs but will not be able to manipulate any of the inner components directly.

This is needed at two levels:
- In the Segment Container
    - This will enable efficient hook-ups for the Table component, which will need to perform frequent Segment Operations.
- In the Storage Writer (add another back-end Segment processor)
    - This will enable async indexing for Table Keys

**Problem location**
Segment Store core

**Suggestions for an improvement**
Add ability to support Container Extensions.
Add ability to support multiple types of StorageWriter processors.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**  
- Added the ability to attach `SegmentContainerExtension`s to `SegmentContainer`s, which have direct access to those Containers and have the same lifecycle as them (they are created when their Container is created and closed when that Container is closed).
- Added the ability to attach `WriterSegmentProcessor`s to the `StorageWriter`, in addition to the main `SegmentAggregator`. These new processors will be served the same data as the main Segment Aggregator, will be flushed periodically, and will be closed when the necessary conditions are met (i.e., at the same time as their sibling Segment Aggregators).

**Purpose of the change**  
Fixes #2854.

**What the code does**  
- Defined a contract for extensions (`SegmentContainerExtension`) and for writer processors (`WriterSegmentProcessor`). Wired them through the various container and writer factories. Added unit tests (for the wiring).
- Renamed `writer.OperationProcessor` to `WriterSegmentProcessor` and moved up one package (in the `server` root. The rename also helps resolve naming collisions with the other `OperationProcessor`.
- Renamed `writer.FlushResult` to `WriterFlushResult` and moved up one package.

The rename and move of the two classes is the reason why so many files are touched - I had to update the imports for all of them. The most interesting part of the whole PR is in the `StorageWriter.java` and `StreamSegmentContainer.java` classes (and their corresponding unit test classes). The rest is just refactoring work.

These features are not yet used. Tables indexing will make use of them and until then they will sit idle and not be exercised by any component.

**How to verify it**  
Unit tests to verify wiring have been added.

### CODE REVIEW

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainer.java [hunk 1]
```diff
@@ -25,4 +24,13 @@
...
     Collection<SegmentProperties> getActiveSegments();
+
+    /**
+     * Gets a registered SegmentContainerPlugin of the given type.
+     *
+     * @param pluginClass Class of the plugin.
+     * @param <T>         Type of the plugin.
+     * @return A registered SegmentContainerPlugin of the requested type.
+     */
+    <T extends SegmentContainerPlugin> T getPlugin(Class<T> pluginClass);
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainer.java]: Isn't the idea the plugin is running internally to the container? If so vending it to the outside seems to break encapsulation. Why is this needed?

Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainer.java]: FYI: at @fpj 's suggestion I renamed `SegmentContainerPlugin` to `SegmentContainerExtension` to make it clear this is not something pluggable via some configuration or dynamically at runtime.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java [hunk 2]
```diff
@@ -124,7 +133,7 @@
         this.durableLog = durableLogFactory.createDurableLog(this.metadata, this.readIndex);
         shutdownWhenStopped(this.durableLog, "DurableLog");
         this.attributeIndex = attributeIndexFactory.createContainerAttributeIndex(this.metadata, this.storage, this.durableLog);
-        this.writer = writerFactory.createWriter(this.metadata, this.durableLog, this.readIndex, this.attributeIndex, this.storage);
+        this.writer = writerFactory.createWriter(this.metadata, this.durableLog, this.readIndex, this.attributeIndex, this.storage, this::createWriterProcessors);
         shutdownWhenStopped(this.writer, "Writer");
```
Comment (RaulGracia) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: I see that in this constructor `createWriterProcessors` is invoked, which internally iterates over `this.extensions`. Then, few lines after in the constructor `this.extensions` is initialized. Is this correct?

Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: It's not invoked, we pass the method reference as a lambda (look at the `::`). That will be invoked when the Writer itself is initialized (when `doStart` on this class is called), which is guaranteed to be after the constructor is done.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java [hunk 3]
```diff
@@ -105,10 +111,13 @@
...
      * @param writerFactory            The WriterFactory to use to create Writers.
      * @param storageFactory           The StorageFactory to use to create Storage Adapters.
+     * @param createExtensions            A Function that, given an instance of this class, will create the set of
+     *                                 {@link SegmentContainerExtension}s to be associated with that instance.
      * @param executor                 An Executor that can be used to run async tasks.
      */
     StreamSegmentContainer(int streamSegmentContainerId, ContainerConfig config, OperationLogFactory durableLogFactory, ReadIndexFactory readIndexFactory,
-                           AttributeIndexFactory attributeIndexFactory, WriterFactory writerFactory, StorageFactory storageFactory, ScheduledExecutorService executor) {
+                           AttributeIndexFactory attributeIndexFactory, WriterFactory writerFactory, StorageFactory storageFactory,
+                           SegmentContainerFactory.CreateExtensions createExtensions, ScheduledExecutorService executor) {
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: One clarification that we might want to make in the documentation (e.g., javadocs) is that these extensions are loaded at construction time (I think that's what you mean with "follows the same lifecycle of the container") and not dynamically. I can't load an extension while the container is running, if I get it right.

One thing I'm wondering about is whether it is possible that an extension is loaded in an segment store instance and not in another. Say that the container changes ownership and the new owner is not running the extension. Is this a valid scenario even?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    void close();
+
+    /**
+     * Gets a value indicating whether the SegmentAggregator is closed (for any kind of operations).
+     */
+    boolean isClosed();
+
+    /**
+     * Gets the SequenceNumber of the first operation that is not fully committed to Storage.
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: Could we add some more detail to the javadoc of this method and others in this interface? For example, define sequence number and explain what property we are trying to preserve by exposing this information to the `StorageWriter`.

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/StreamSegmentContainerTests.java [hunk 5]
```diff
@@ -1513,7 +1521,60 @@ public void testStartOffline() throws Exception {
...
+        val p = context.container.getExtension(TestSegmentContainerExtension.class);
+        Assert.assertEquals("Unexpected result from getExtension().", extension.get(), p);
+
+        // Verify Writer Segment Processors are properly wired in.
+        String segmentName = getSegmentName(0);
+        byte[] data = getAppendData(segmentName, 0);
+        context.container.createStreamSegment(segmentName, null, TIMEOUT).join();
+        context.container.append(segmentName, data, null, TIMEOUT).join();
+        val rawOp = operationProcessed.get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+        Assert.assertTrue("Unexpected operation type.", rawOp instanceof CachedStreamSegmentAppendOperation);
```
Comment (fpj) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/StreamSegmentContainerTests.java]: I don't actually understand how we are checking that the segment processors are properly wired in. Is this only checking that they are not creating any deviation of the correct behavior?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterFactory.java [hunk 6]
```diff
@@ -25,8 +26,21 @@
...
+     * @param createProcessors  A Function that, when invoked with a Segment's Metadata, will create any additional
+     *                          WriterSegmentProcessors for that Segment.
      * @return An instance of a class that implements the Writer interface.
      */
     Writer createWriter(UpdateableContainerMetadata containerMetadata, OperationLog operationLog, ReadIndex readIndex,
-                        ContainerAttributeIndex attributeIndex, Storage storage);
+                        ContainerAttributeIndex attributeIndex, Storage storage, CreateProcessors createProcessors);
+
+    @FunctionalInterface
+    interface CreateProcessors {
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterFactory.java]: I'm curious to know how we will be adding extensions that we develop over time. Is there going to be an implementation of this interface somewhere that we add to?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainer.java [hunk 7]
```diff
@@ -25,4 +24,13 @@
...
     Collection<SegmentProperties> getActiveSegments();
+
+    /**
+     * Gets a registered {@link SegmentContainerExtension} of the given type.
+     *
+     * @param extensionClass Class of the {@link SegmentContainerExtension}.
+     * @param <T>         Type of the {@link SegmentContainerExtension}.
+     * @return A registered {@link SegmentContainerExtension} of the requested type.
+     */
+    <T extends SegmentContainerExtension> T getExtension(Class<T> extensionClass);
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainer.java]: Yes. This should be more than sufficient for now.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerExtension.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+     * @return A CompletableFuture that, when completed, will indicate that the initialization is done.
+     */
+    CompletableFuture<Void> initialize();
+
+    /**
+     * Creates a Collection of any additional WriterSegmentProcessors to pass to the StorageWriter.
+     * @param metadata The Metadata of the Segment to create Writer Processors for.
+     * @return A Collection containing the desired processors. If none are required, this should return an empty list.
+     */
+    Collection<WriterSegmentProcessor> createWriterSegmentProcessors(UpdateableSegmentMetadata metadata);
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerExtension.java]: I need a way to inject "extensions" into the `StorageWriter` as well (at the segment level). These extensions are associated with the `SegmentContainerExtension` that owns them and has the context required to create them. 

The alternative to the presented approach would be to have another type of extension that is passed to the `StorageWriterFactory` (different from `SegmentContainerFactory`) so that it can create that. That will not only lead to a lot more code and possibilities of messing up (like setting up one but forgetting the other one), but I will not be able to communicate between the `SegmentContainerExtension` and the `WriterTableProcessor`s associated with it. For Tables, I need the former to know when the latter has finished post-recovery indexing so that it may serve conditional updates.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java (hunk 3 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: The "loading" is hard coded in `ServiceBuilder` and is not driven by configuration. So as long as all Segment Store instances run the same code, we are fine. If they don't, we have much bigger problems to worry about.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java (hunk 4 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: Sequence Number is universally known in the Segment Container and all operations have it. This is not overriding its definition hence it does not need extra clarification.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterFactory.java (hunk 6 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterFactory.java]: This is a functional interface. It defines a lambda function signature that can be passed in. This is internal; you don't need anything extra. 

The way you add additional extensions is by registering them in `ServiceBuilder`. You will see in a subsequent PR which depends on this one.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerExtension.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines an Extension that can be associated with a {@link SegmentContainer}, which will be instantiated (and initialized)
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainerExtension.java]: ok

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java (hunk 3 above)
Comment (fpj) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: How are upgrades supposed to work? Say I introduce an extension in a given version _X_ and I want to upgrade my Pravega cluster from version _X - 1_ to _X_. My cluster has data already and I don't want to wipe it out. Have you thought about it or do we need to figure this out?

Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: This has nothing to do with data as it doesn't manipulate anything that the container stores or accesses. Extensions only use external container APIs and they have no access to internals.

Take for example the Tables extension: it doesn't touch the internals of the container but it add another layer on top of that that allows us to expose data in specific segments in a specified format (dictated by the extension); nothing is changed underneath that.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java (hunk 4 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: I really do not like to copy paste comments for the same reason why I don't like to copy-paste code. Once you change it in one place you must remember to change it everywhere else, otherwise any number of bad things can happen. If I define SNs in the container's Javadoc, then define them again here, if I find myself needing to alter the description in the container's Javadoc then I must remember that I also wrote it here too. If I don't, then someone will likely get confused and end up doing something they shouldn't do.

This is not an external API on top of a black/grey box. It's an internal contract so anyone who wishes to write this will likely have access to the full code of the Segment Store core, and will need to deeply understand how it works. Sequence Numbers are one of those building blocks that are needed to comprehend the behavior of anything this low in the API stack.

Comment (fpj) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: Note that I also said that you could add a pointer, I understand the problems involved in duplicating definitions. Up to you, I think it would improve clarity.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Defines a general Segment Processor for the Storage Writer.
+ */
+public interface WriterSegmentProcessor extends AutoCloseable {
+
+    @Override
+    void close();
+
+    /**
+     * Gets a value indicating whether the SegmentAggregator is closed (for any kind of operations).
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: This comment is referring to `SegmentAggregator`, but this is not about the segment aggregator alone.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriter.java [hunk 11]
```diff
@@ -519,4 +537,103 @@ public String toString() {
...
+    //region ProcessorCollection
+
+    /**
+     * Wraps a collection of WriterSegmentProcessors, including the main Segment Aggregator.
+     */
+    private static class ProcessorCollection implements WriterSegmentProcessor {
+        private final SegmentAggregator aggregator;
+        private final List<WriterSegmentProcessor> processors;
+
+        ProcessorCollection(SegmentAggregator aggregator, Collection<WriterSegmentProcessor> processors) {
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriter.java]: I don't understand why you need to separate out the aggregator. Is it convenience to avoid having to look for it in the collection? Since this is internal, I don't really care, but I'd like to at least have a clear justification as a response to this comment, please.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+ */
+package io.pravega.segmentstore.server;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines a general Segment Processor for the Storage Writer.
+ */
+public interface WriterSegmentProcessor extends AutoCloseable {
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: This interface has no call defined indicating processing. Given that it is the contract of a `Processor`, shouldn't it have a method that indicates processing?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java (hunk 3 above)
Comment (fpj) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: I see your point, but I'm still wondering about rolling upgrades when adding an extension based on this comment:

> as long as all Segment Store instances run the same code, we are fine

which indicates that we can't do a rolling upgrade of the segment store when upgrading to a version that has a new extension. Maybe that's ok because a feature can't really be exercised when only some instances have it. Take tables as an example. An application can't really use when some segment store instances are missing the new code. Either we have a way to spot when all instances are running it to enable the extension or we require a planned outage. Also, extensions are expected to be added infrequently, so we can plan on which versions to introduce them.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriter.java (hunk 11 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriter.java]: Yes, it's a convenience so I don't have to always access the first element in the list when I need it.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java (hunk 12 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/WriterSegmentProcessor.java]: It has two methods:
1. `add(Operation)`
2. `flush()`

Together, they perform some sort of processing.

**Comment by andreipaduroiu** [comment]
@fpj 
https://github.com/pravega/pravega/pull/2861#discussion_r220364619

You are referring to *adding* a new feature. In order to make use of it, no client should make use of it before all servers can support it. Similarly, for deprecation, all consuming code must not use that feature and then it can be removed from the server code. That should be an easy thing to do. Why would you deploy your client code that makes use of a new feature when that feature isn't available on the server? I haven't seen any client-server app that can pull any magic tricks here; all of them will throw back some sort of an operation-not-supported exception if you try to do that.

As for upgrading an existing feature, we've already been through that exercise with versioning earlier this year.

This case is an example of the former, not the latter.
